### PR TITLE
Comparison tool

### PR DESCRIPTION
**Pull request by kakoeh**
New cli tool "compare" ([odxtools - Issue 35](https://github.com/mercedes-benz/odxtools/issues/35))
Can be used to compare databases (pdx-files) and/or ecu variants with each other.
Compares all diagnostic services as well as its parameters of the specified databases /variants with each other and displays the changes.

Katja Köhler &lt;katja.koehler@mercedes-benz.com&gt;, on behalf of [MBition GmbH](https://mbition.io/).
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

### CODE REVIEW

File: odxtools/cli/find.py [hunk 1]
```diff
@@ -120,6 +121,14 @@ def add_subparser(subparsers: "argparse._SubParsersAction") -> None:
         help="Relax output formatting rules (allow unknown bitlengths for ascii representation)",
     )
 
+    parser.add_argument(
+        "-od",
+        "--object-details",
+        action="store_true",
+        required=False,
+        help="Print all object details instead of just the name",
```
Comment (kayoub5) [inline odxtools/cli/find.py:129]: this new argument seems unrelated to the compare function, so it should be moved to its own PR along with any other code that was added for it

File: odxtools/database.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 from typing import List, Optional
 from xml.etree import ElementTree
 from zipfile import ZipFile
+
 from packaging.version import Version
```
Comment (kayoub5) [inline odxtools/database.py:8]: meaningless change

Comment (kakoeh) [reply odxtools/database.py:8]: Caused by reformat-source script

File: odxtools/cli/find.py (hunk 1 above)
Comment (kakoeh) [reply odxtools/cli/find.py:129]: The _compare_ function uses the _print_service_parameters_ function to display information about the service parameters. The _find_ function uses the _print_diagnostic_service_ function to display diagnostic service details. These util-functions were partially identical in their functionality, so the _print_diagnostic_service_ function has been adapted and now calls the _print_service_parameters_ function if necessary. As the _print_service_parameters_ function has an argument specifying whether the entire diagnostic object should be displayed or not, the _find_ function needed a new argument.

File: odxtools/cli/list.py [hunk 3]
```diff
@@ -178,6 +193,14 @@ def add_subparser(subparsers: "argparse._SubParsersAction") -> None:
         help="Print a list of all diagnostic services and DOPs specified in the pdx",
     )
 
+    parser.add_argument(
```
[the remark this answers is not in the document]
Comment (kakoeh) [reply odxtools/cli/list.py:197]: The _compare_ function uses the _print_service_parameters_ function to display information about the service parameters. The _list_ function uses the _print_diagnostic_service_ function to display diagnostic service details. These util-functions were partially identical in their functionality, so the _print_diagnostic_service_ function has been adapted and now calls the _print_service_parameters_ function if necessary. As the _print_service_parameters_ function has an argument specifying whether the entire diagnostic object should be displayed or not, the _list_ function needed a new argument.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -61,6 +61,7 @@ jobs:
           # stumbles over. for some reason, mypy cannot be
           # instructed to ignore 'version.py'
           sed -i "s/# type:.*//" odxtools/version.py
+          python -m mypy --install-types
```
Comment (kayoub5) [inline .github/workflows/tests.yml:64]: Changes to CI can not be mixed to other changes such as adding a new feature

File: odxtools/cli/find.py (hunk 1 above)
Comment (kayoub5) [reply odxtools/cli/find.py:129]: The split this PR in two then:
* One PR to add print_objects to print_diagnostic_service
* One PR to add compare tool, and use the print_diagnostic_service

Any other way of splitting this PR is also acceptable, but the PR as is with +1400 changes, is too big for review

See https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/

**Comment by kayoub5** [comment]
Where did the file `examples/somersault_01.01.02.pdx` come from?

File: pyproject.toml [hunk 5]
```diff
@@ -42,6 +42,7 @@ dependencies = [
     "markdownify >= 0.11",
     "deprecation >= 2.1",
     "packaging",
+    "pandas >= 2.1.4"
```
Comment (kayoub5) [inline pyproject.toml:45]: pandas is a test dependency or library dependency?

File: .github/workflows/tests.yml (hunk 4 above)
Comment (kakoeh) [reply .github/workflows/tests.yml]: Outdated, missing packages for CI test were added to pyproject.toml

File: pyproject.toml (hunk 5 above)
Comment (kakoeh) [reply pyproject.toml]: > Where did the file `examples/somersault_01.01.02.pdx` come from?

The file was derived from the file `somersault.pdx` by adapting `somersaultecu.py`

> pandas is a test dependency or library dependency?

pandas was used by several functions but has been replaced

File: odxtools/cli/_print_utils.py [hunk 6]
```diff
@@ -53,31 +75,206 @@ def print_diagnostic_service(
...
                                                      len(service.negative_responses) > 1):
         # Does this ever happen?
         raise NotImplementedError(
             f"The diagnostic service {service.odx_id} offers more than one response!")
+
+
+def generatetable(style: str, list_of_items: Union[List[DiagService],
+                                                   List[Parameter]]) -> pd.DataFrame:
+    # generates table for different properties of diagnostic service or parameter of diagnostic service
+    # style = {'service-details', 'parameter-details'}
```
[the remark this answers is not in the document]
Comment (kakoeh) [reply odxtools/cli/_print_utils.py:152]: Comment shows the set of possible arguments

[the remark this answers is not in the document]
Comment (kayoub5) [reply odxtools/cli/_print_utils.py:152]: Use doc string format then (triple quotes)

File: pyproject.toml (hunk 5 above)
Comment (kayoub5) [reply pyproject.toml]: what are the adaptations?

@andlaus are you ok with adding a binary file without the logic that created it?

Comment (kakoeh) [reply pyproject.toml]: - new ecu variant `somersault_young`
- deleted SingleEcuJob `compulsory_program`
- deleted DiagService `set_operation_params`
- new DiagService `flicflac`
- renamed DiagService `start_session` from `session_start` to `start_session`
- changed pos. response parameter of `session_start`
(DOP of `can_do_backward_flips` from `uint8` to `boolean`)
- changed pos. response parameter of `tester_present`
(value of `status` from `0x00` to `0x0043` and type from `uint8` to `uint16`)
- changed pos. response parameter of `grudging_backward`
(byte position of `num_flips_done` from `None` to `1`)
(param type of `grumpiness_level` from `ValueParameter` to `CodedConstParameter`, 
incl. byte position from `None` to `2`, value = `0x05`)
- changed pos. response parameter of `status_report`
(byte position of `happiness_level` from `2` to `1`)
(byte position of `dizzyness_level` from `1` to `2`)
- changed request parameter of `stop_session`
(type of `id` from `uint8` to `uint16`)
- changed request parameter of `tester_present`
(value of `id` from `0x0` to `0x01`)
- changed request parameter of `do_backward_flips`
(short_name of `backward_check` from `backward_soberness_check` to `backward_check`)

image

File: odxtools/cli/find.py (hunk 1 above)
Comment (kakoeh) [reply odxtools/cli/find.py:129]: Another PR for functionalities unrelated to compare function was created: #250

File: odxtools/cli/list.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kakoeh) [reply odxtools/cli/list.py:201]: See above

File: pyproject.toml (hunk 5 above)
Comment (andlaus) [reply pyproject.toml]: > @andlaus are you ok with adding a binary file without the logic that created it?

I suppose in this case it is okay, because that new file is basically a showcase for the `compare` tool. (I'd prefer to have a script doing these modifications in-repo, though.)

can this file be renamed to something which better conveys the fact that it is a modified file of the original? (something like `somersault_modified.pdx`.)

File: odxtools/database.py (hunk 2 above)
Comment (kayoub5) [reply odxtools/database.py:8]: @andlaus most likely caused by an upgrade (new version) of yapf

File: odxtools/cli/compare.py [hunk 7]
```diff
@@ -0,0 +1,730 @@
...
+
+    parser.add_argument(
+        "-po",
+        "--plumbing-output",
+        action="store_true",
+        required=False,
+        help="Print full objects instead of selected and formatted attributes",
+    )
+    # TODO
+    # Idea: provide folder with multiple pdx files as argument
```
Comment (kayoub5) [inline odxtools/cli/compare.py:601]: maybe create an issue? otherwise this will get lost in all the code

Comment (kakoeh) [reply odxtools/cli/compare.py:601]: Ideas for improvement of comparison tool: https://github.com/mercedes-benz/odxtools/issues/252

**Comment by andlaus** [comment]
nice, thanks all!
